### DISCUSSION TITLE
Failed build image by packeto-buldpacks jdk

### DISCUSSION THREAD

**Discussion by heesuk-ahn** [paketo-buildpacks/feedback#29] [Q&A] [answered]
Hi, All

I set the JDK of packeto buildpack in spring boot and this is it.

```
bootBuildImage {
    buildpacks = ["gcr.io/paketo-buildpacks/adoptium:latest", "gcr.io/paketo-buildpacks/java:latest"]
}
```

But recently, it suddenly started failing during the build process.

Error Bellow:

```
Invalid response received when loading image "pack.local/builder/iabgkkeeuy:latest"
```

If you do not directly inject the buildpack JDK, it works normally. (default liberica)
This is the same situation in `CI SERVER` as well as my `Local PC`

Also,

I built it using the pack cli. In this case, an error like this occurs.

```
command :  pack build build-test --path . --buildpack paketo-buildpacks/ca-certificates --buildpack paketo-buildpacks/azul-zulu 
--buildpack paketo-buildpacks/java
```

```
paketo-buildpacks/java
base: Pulling from paketobuildpacks/builder
Digest: sha256:3c5369193865a530cd37b7c86d48fa72ca19f879cb14b5f6aaee671952cfe451
Status: Image is up to date for paketobuildpacks/builder:base
base-cnb: Pulling from paketobuildpacks/run
Digest: sha256:0600ef2dd08ea48fa5d4423e18e4efb1039ff482e5afd3ca33d6993a10079f4b
Status: Image is up to date for paketobuildpacks/run:base-cnb
ERROR: failed to build: downloading buildpack: extracting from registry paketo-buildpacks/azul-zulu: fetching image: resolve auth for ref gcr.io/paketo-buildpacks/azul-zulu@sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf: error getting credentials - err: exec: "docker-credential-gcloud": executable file not found in $PATH, out: ``
```

**Comment by dmikusa** [comment]
That doesn't sound like a Buildpacks issue. Maybe a missing dependency.
What  OS are you using? What version of pack?

**Reply by heesuk-ahn** [reply]
My Os is Mac - `Monterey 12.3 version`
and my pack version is `0.26.0+git-04f7318.build-3309` and Docker version 20.10.13

**Reply by heesuk-ahn** [reply]
The spring boot gradle plguin - `bootBuildImage` task also fails to build the image. 🤔 
The same thing happens on the ci server and on someone else's local PC.

**Reply by dmikusa** [reply]
Are you using Docker Desktop or some other Docker implementation on your Mac? i.e. Rancher Desktop, minikube, plain-old Linux VM with Docker, etc...

**Reply by heesuk-ahn** [reply]
I am using Docker Desktop. (version 4.6.1)

**Reply by dmikusa** [reply]
Also, can you `docker pull` the image that's failing?

```
> docker pull gcr.io/paketo-buildpacks/azul-zulu@sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf
gcr.io/paketo-buildpacks/azul-zulu@sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf: Pulling from paketo-buildpacks/azul-zulu
0319d711b718: Pull complete
Digest: sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf
Status: Downloaded newer image for gcr.io/paketo-buildpacks/azul-zulu@sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf
gcr.io/paketo-buildpacks/azul-zulu@sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf
```

**Reply by heesuk-ahn** [reply]
One strange thing is that there is no change in Spring Boot version or code level or docker version, but all bootBuildImage Tasks are broken. 🤔

**Reply by heesuk-ahn** [reply]
```
docker pull gcr.io/paketo-buildpacks/azul-zulu@sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf
gcr.io/paketo-buildpacks/azul-zulu@sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf: Pulling from paketo-buildpacks/azul-zulu
0319d711b718: Pull complete 
Digest: sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf
Status: Downloaded newer image for gcr.io/paketo-buildpacks/azul-zulu@sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf
gcr.io/paketo-buildpacks/azul-zulu@sha256:42754e8954619b0dc022ab69f176c75f29009a7a16463b415f79196a924a95cf
```

docker pull succeeds.

**Reply by dmikusa** [reply]
It seems like a problem downloading the image. `docker-credential-gcloud` is a binary that docker will call to perform auth for google cloud, the image in question is on `gcr.io`. I don't know why it's trying to auth though because that's a public image. It shouldn't require any auth.

**Reply by heesuk-ahn** [reply]
The strange thing is that if we use Default without setting the buildPacks JDK directly, the image build succeeds normally.

```
// build failed
bootBuildImage {
    buildpacks = ["gcr.io/paketo-buildpacks/adoptium:latest", "gcr.io/paketo-buildpacks/java:latest"]
    environment("BP_DATADOG_ENABLED", "true")
}
```

```
// build success
bootBuildImage {
    environment("BP_DATADOG_ENABLED", "true")
}
```

**Reply by heesuk-ahn** [reply]
https://github.com/spring-projects/spring-boot/blob/4bb5ba2bd0609e72bfdcdd2f815c4085e5e9584d/spring-boot-project/spring-boot-tools/spring-boot-buildpack-platform/src/main/java/org/springframework/boot/buildpack/platform/docker/DockerApi.java#L230-L250

This is probably because the image was not created normally and the Load failed.

**Reply by dmikusa** [reply]
I'm not able to replicate the issue, sorry. Other thoughts...

1. You might try from a different network? Perhaps if it's a corp networking issue, like if you're behind a proxy or something it's causing problems.
2. If you have an internal Registry (like Harbor or Docker's registry), you could cache the buildpacks all there. You can then point the tools to pull buildpack from your internal registry. You can use `pack config registry-mirrors -h` to set a registry mirror that pack will use (Spring Boot doesn't have the equivalent at the moment). You can also just set your `pack build -b` arguments or buildpacks in Maven/Gradle config to point to `your-registry.example.com/paketo-buildpacks/...` and it'll pull the images from there. You can use `crane cp` to get the images to your registry pretty easily.
3. I don't think it should need it but, install "docker-credential-gcloud" (or make sure it's on the $PATH) so that it can be found. I think the google cloud SDK provides this. Again, it shouldn't be necessary but might be interesting to see what happens if that is present.

**Reply by dmikusa** [reply]
>The strange thing is that if we use Default without setting the buildPacks JDK directly, the image build succeeds normally.

That is very likely because in that case it has all of the buildpacks already. They are present in the builder image, which you already have downloaded, so it doesn't need to go to the Internet/gcr.io.

**Reply by heesuk-ahn** [reply]
Thanks for the good suggestions. The same problem occurs when not using a VPN, but I will test it after registering it in the company registry first. :) and then I will share result later

**Reply by backjo** [reply]
This was separately reported on https://github.com/paketo-buildpacks/base-builder/issues/473. Just wanted to add that we're independent from these two reporters and also seeing this issue. Setting builder = "paketobuildpacks/builder:0.2.69-base" in the gradle spring boot plugin configuration is a workaround that worked for us. 

@dmikusa-pivotal  - This is minimally reproducible using the Spring Boot Initializr with a slightly modified configuration -> https://github.com/backjo/broken-buildpack-example . I've setup a Github Action CI run on the repository to demonstrate the error  reproducing -> https://github.com/backjo/broken-buildpack-example/runs/6701052679?check_suite_focus=true

**Reply by dmikusa** [reply]
@backjo I see that your app demo is failing, but there's no clear indication of why (beyond `Invalid response received when loading image`, which isn't that helpful).

Are you able to reproduce this with `pack`? `pack -v` is often more helpful when debugging issues like this. I tried but couldn't get it to reproduce with pack. That would also help to narrow down if the issue you're seeing is a problem specific to the Maven/Gradle integration.

**Reply by heesuk-ahn** [reply]
@backjo 

I tried your solution about setting builder version.
so, it's worked for me 👍 

```
bootBuildImage {
    builder = "paketobuildpacks/builder:0.2.69-base"
    buildpacks = ["gcr.io/paketo-buildpacks/adoptium:latest", "gcr.io/paketo-buildpacks/java:latest"]
}
```

**Reply by backjo** [reply]
Still troubleshooting, but it looks like the underlying cause is 2 additional new Docker layers in 0.2.70-base and beyond that are causing the Spring Boot plugins implementation to hit the 127 limit.

**Comment by scottfrederick** [comment]
There is an explanation of the issue and a few possible work-arounds here: https://github.com/spring-projects/spring-boot/issues/31233#issuecomment-1145220375.

**Comment by dmikusa** [comment, marked as answer]
OK, so there are two issues being reported here:

1. Running Maven/Gradle to build an image with buildpacks and it fails. That is addressed by @scottfrederick's comment: https://github.com/orgs/paketo-buildpacks/discussions/29#discussioncomment-2873763

2. `err: exec: "docker-credential-gcloud": executable file not found in $PATH, out: `. This error can happen for a few reasons, but it is a local workstation problem. The pack cli will delegate to Docker for authentication, so if you have an entry in your `docker.config` file for `gcr.io` it will attempt to use that to authenticate, even if the image it's loading is public. The pack cli doesn't not check if it's public first. This error often happens if you had gcloud client utilities installed and removed them or the install got messed up (sometimes happens on MacOS after doing a system update). There's a few possible fixes: a.) if you don't use gcloud, edit your `docker.config` file and remove the entry for `gcr.io` b.) If you use gcloud but removed the tools, install or reinstall them c.) if they are already installed, you may need to reinstall the gcloud credentials helper again. At any rate, once you sort this out `pack` should start working again. [More background here](https://github.com/buildpacks/pack/issues/728#issuecomment-655656998).
